src/main/scala/com/daml/projection/JdbcAction.scala

```diff
@@ -162,42 +166,69 @@ object Sql {
-      // TODO fix in typeclass solution
-      case x: Option[_] if x.isEmpty => ps.setNull(pos, java.sql.Types.NULL)
-      case Some(x)                   => ps.setObject(pos, x)
-      case x                         => ps.setObject(pos, x)
-    }
-  }
+@FunctionalInterface
+trait Bind[-T] {
+
+  /** Sets the value on the [[java.sql.PreparedStatement]] */
+  def set(ps: PreparedStatement, pos: Int, value: T): Unit
+
+  /** Creates a [[Bind.Applied]] function that will set `value` at `pos` on a `PreparedStatement` argument. */
+  final def apply(value: T, pos: Int): Bind.Applied = set(_, pos, value)
 }
 
-/** Sets a bind value on a PreparedStatement */
-@SerialVersionUID(1L)
-@FunctionalInterface
-trait Setter extends java.io.Serializable {
+/**
+ * Provides `Bind` values for [[ExecuteUpdate.bind[T](pos*]],[[ExecuteUpdate.bind[T](name*]], [[Binder.bind[T](pos*]]
+ * and [[Binder.bind[T](name*]] methods.
+ */
+object Bind {
+  type Used[-T] = Bind[_ >: T]
+
+  private def mk[T <: AnyVal](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = f(ps, pos, value)
+  }
+
+  private def mkN[T >: Null <: AnyRef](f: (PreparedStatement, Int, T) => Unit): Bind[T] = new Bind[T] {
+    override def set(ps: PreparedStatement, pos: Int, value: T) = {
+      if (value eq null) ps.setNull(pos, java.sql.Types.NULL) else f(ps, pos, value)
+    }
+    // TODO SC would a specific thing from Types be better?
+  }
 
-  /** Sets a bind value on `ps` */
-  def set(ps: java.sql.PreparedStatement): Unit
+  implicit val Boolean: Bind[Boolean] = mk(_.setBoolean(_, _))
+  implicit val Byte: Bind[Byte] = mk(_.setByte(_, _))
+  implicit val Short: Bind[Short] = mk(_.setShort(_, _))
+  implicit val Int: Bind[Int] = mk(_.setInt(_, _))
+  implicit val Long: Bind[Long] = mk(_.setLong(_, _))
+  implicit val Float: Bind[Float] = mk(_.setFloat(_, _))
+  implicit val Double: Bind[Double] = mk(_.setDouble(_, _))
+  implicit val `scala BigDecimal`: Bind[BigDecimal] = mkN((ps, pos, x) => ps.setBigDecimal(pos, x.bigDecimal))
+  implicit val BigDecimal: Bind[java.math.BigDecimal] = mkN(_.setBigDecimal(_, _))
+  implicit val String: Bind[String] = mkN(_.setString(_, _))
+  implicit val Date: Bind[java.sql.Date] = mkN(_.setDate(_, _))
+  implicit val LocalDate: Bind[LocalDate] = mkN((ps, pos, x) => ps.setDate(pos, java.sql.Date.valueOf(x)))
+  implicit val Timestamp: Bind[java.sql.Timestamp] = mkN(_.setTimestamp(_, _))
+  implicit val LocalDateTime: Bind[LocalDateTime] =
+    mkN((ps, pos, x) => ps.setTimestamp(pos, java.sql.Timestamp.valueOf(x)))
+  implicit val Instant: Bind[Instant] =
+    mkN((ps, pos, x) => ps.setTimestamp(pos, new java.sql.Timestamp(x.toEpochMilli)))
+  implicit val Array: Bind[java.sql.Array] = mkN(_.setArray(_, _))
+  implicit val Null: Bind[Null] = mkN((ps, pos, _) => ps.setNull(pos, java.sql.Types.NULL))
+  // TODO SC is Types.NULL right?
```

Comment:
This was here in case someone passes `null` without a type. I think we can probably do away with that, or at least do away with its `implicit` status; if we want to keep it we should have a Scala test that checks that implicit resolution uses this in the correct case (it needs to move to a lower priority position if we can't).

Reply:
I'll remove it.